modules/python/python-fetcher.c

```diff
@@ -0,0 +1,470 @@
+  g_hash_table_insert(self->options, normalized_key, g_strdup(value));
+}
+
+void
+python_fetcher_set_loaders(LogDriver *s, GList *loaders)
+{
+  PythonFetcherDriver *self = (PythonFetcherDriver *) s;
+
+  string_list_free(self->loaders);
+  self->loaders = loaders;
+}
+
+static const gchar *
+python_fetcher_format_stats_instance(LogThreadedSourceDriver *s)
+{
+  PythonFetcherDriver *self = (PythonFetcherDriver *) s;
+  static gchar persist_name[1024];
+
+  if (s->super.super.super.persist_name)
+    g_snprintf(persist_name, sizeof(persist_name), "python-fetcher,%s", s->super.super.super.persist_name);
+  else
+    g_snprintf(persist_name, sizeof(persist_name), "python-fetcher,%s", self->class);
+
+  return persist_name;
+}
+
+static void
+_pf_py_invoke_void_method_by_name(PythonFetcherDriver *self, const gchar *method_name)
+{
+  _py_invoke_void_method_by_name(self->py.instance, method_name, self->class, self->super.super.super.super.id);
+}
+
+static gboolean
+_pf_py_invoke_bool_method_by_name_with_args(PythonFetcherDriver *self, const gchar *method_name)
+{
+  return _py_invoke_bool_method_by_name_with_args(self->py.instance, method_name, self->options, self->class,
+                                                  self->super.super.super.super.id);
+}
+
+static void
+_pf_py_invoke_void_function(PythonFetcherDriver *self, PyObject *func, PyObject *arg)
+{
+  return _py_invoke_void_function(func, arg, self->class, self->super.super.super.super.id);
+}
+
+static gboolean
+_pf_py_invoke_bool_function(PythonFetcherDriver *self, PyObject *func, PyObject *arg)
+{
+  return _py_invoke_bool_function(func, arg, self->class, self->super.super.super.super.id);
+}
+
+static gboolean
+_py_invoke_init(PythonFetcherDriver *self)
+{
+  return _pf_py_invoke_bool_method_by_name_with_args(self, "init");
+}
+
+static void
+_py_invoke_deinit(PythonFetcherDriver *self)
+{
+  _pf_py_invoke_void_method_by_name(self, "deinit");
+}
+
+static void
+_py_invoke_request_exit(PythonFetcherDriver *self)
+{
+  _pf_py_invoke_void_function(self, self->py.request_exit_method, NULL);
+}
+
+static gboolean
+_py_invoke_open(PythonFetcherDriver *self)
+{
+  return _pf_py_invoke_bool_function(self, self->py.open_method, NULL);
+}
+
+static void
+_py_invoke_close(PythonFetcherDriver *self)
+{
+  _pf_py_invoke_void_function(self, self->py.close_method, NULL);
+}
+
+static inline gboolean
+_ulong_to_fetch_result(unsigned long ulong, ThreadedFetchResult *result)
+{
+  switch (ulong)
+    {
+    case THREADED_FETCH_ERROR:
+    case THREADED_FETCH_NOT_CONNECTED:
+    case THREADED_FETCH_SUCCESS:
+      *result = (ThreadedFetchResult) ulong;
+      return TRUE;
+
+    default:
+      return FALSE;
+    }
+}
+
+static LogThreadedFetchResult
+_py_invoke_fetch(PythonFetcherDriver *self)
+{
+  PyObject *ret = _py_invoke_function(self->py.fetch_method, NULL, self->class, self->super.super.super.super.id);
+
+  if (!ret || !PyTuple_Check(ret) || PyTuple_Size(ret) > 2)
```

Comment:
Why not assert for 2 item? Should we be okay without msg, even if some cases that should be always a dummy NULL?